### ISSUE TITLE
Missing override checks lead to unsoundness

### ISSUE THREAD

**Issue by smarter** [scala/bug#11042]
```scala
trait Base[+A] {
  def get: A = null.asInstanceOf[A]
}

trait SubBase extends Base[Any] {
  override def get: Any = ""
}

class Hi extends Base[Int] with SubBase

object Test {
  def main(args: Array[String]): Unit = {
    val hi = new Hi
    val base: Base[Int] = hi
    val y: Int = base.get // ClassCastException: java.lang.String cannot be cast to java.lang.Integer
    println(y)
  }
}
```
If we make `get` in `Base` abstract, we get an error as expected:
```scala
opt.scala:9: error: overriding method get in trait Base of type => Int;
 method get in trait SubBase of type => Any has incompatible type
class Hi extends Base[Int] with SubBase
      ^
one error found
```
Dotty always produces an error:
```scala
-- Error: try/opt.scala:9:6 ----------------------------------------------------
9 |class Hi extends Base[Int] with SubBase
  |      ^
  |method get in trait SubBase is not a legal implementation of `get` in class Hi
  |  its type             => Any
  |  does not conform to  => Int
one error found
```

This is somewhat important since it affects the design of the collections, see https://github.com/lampepfl/dotty/issues/4839

**Comment by Blaisorblade** [comment]
FWIW: On our side, the relevant checks were implemented in https://github.com/lampepfl/dotty/pull/4013/files#diff-d9f0cf9c3f67df900a2d657a417633b5R628.

**Comment by adriaanm** [comment]
looks related to #11136

**Comment by adriaanm** [comment]
this one is fixed by my fix for #11136. I'll add a test next week.

```
Welcome to Scala 2.13.0-pre-f86a0d1-local (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_181).
Type in expressions for evaluation. Or try :help.

scala> 

scala> :paste
// Entering paste mode (ctrl-D to finish)

trait Base[+A] {
  def get: A = null.asInstanceOf[A]
}

trait SubBase extends Base[Any] {
  override def get: Any = ""
}

class Hi extends Base[Int] with SubBase

object Test {
  def main(args: Array[String]): Unit = {
    val hi = new Hi
    val base: Base[Int] = hi
    val y: Int = base.get // ClassCastException: java.lang.String cannot be cast to java.lang.Integer
    println(y)
  }
}

// Exiting paste mode, now interpreting.

       class Hi extends Base[Int] with SubBase
             ^
On line 9: error: incompatible type in overriding
       def get: Int (defined in trait Base)
         with override def get: Any (defined in trait SubBase);
        found   : => Any
        required: => Int

```

**Comment by SethTisue** [comment]
closing since the bug is believed fix. missing test is now https://github.com/scala/scala-dev/issues/622
